### PR TITLE
[codex] Harden renderer teardown polish

### LINKED ISSUE #3647: Follow up renderer teardown review polish from #3576
Follow-up from #3576 after merging the renderer-function teardown contract.

Non-blocking review polish to evaluate separately:

- Decide whether to share `invokeRendererTeardown` through an internal OSS export for Pro, or add a focused cross-package test that keeps the duplicated OSS/Pro helpers aligned.
- Add extra test hardening for concurrent async renderer races where a stale async renderer resolves after a newer same-id render has replaced its placeholder.
- Add explicit tests for `wrapServerComponentRenderer` early error messages (`domNodeId`, missing DOM node, missing `railsContext`).
- Consider a stronger runtime shape check in `wrapServerComponentRenderer` for the theoretically unreachable case where a registered server component render function resolves to a teardown-shaped function instead of a React component.
- Hoist duplicated `setupRailsContext` helpers in `ClientRenderer.test.ts` if future test edits touch that file again.
- Revisit the public `RendererFunction` return typing after downstream migration data is available; the broad legacy-compatible return type is intentional in #3576, but new renderer authors could benefit from stricter helpers or docs.
- Keep the long-form docs and dummy-app renderer conversions aligned with #3585 and the follow-up work that remains after #3576 lands.

These are follow-up hardening/documentation tasks. They are not blockers for merging #3576 because the branch already fixes the P2 teardown behavior found in review and CI is green.

### PR DESCRIPTION
**Pull request by justin808**
Closes #3647

## Summary
- Centralize `isRendererTeardownResult` in `packages/react-on-rails/src/rendererTeardown.ts` and expose it as `react-on-rails/@internal/rendererTeardown`, so OSS and Pro share one teardown-result guard instead of carrying inline copies.
- Keep the Pro `wrapServerComponentRenderer/client` runtime guard so teardown-shaped server-component render results fail with a clear error before the generic component-shape check.
- Harden and clarify the renderer teardown tests around same-id async races, DOM descriptor reuse, and Pro wrapper validation cleanup.
- Keep Pro parity coverage for non-native thenable teardown rejection handling.

## Tests
- `pnpm install --frozen-lockfile`
- `pnpm exec prettier --check packages/react-on-rails/src/rendererTeardown.ts packages/react-on-rails/src/ClientRenderer.ts packages/react-on-rails/src/createReactOutput.ts packages/react-on-rails-pro/src/ClientSideRenderer.ts packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx packages/react-on-rails/tests/ClientRenderer.test.ts packages/react-on-rails/package.json`
- `pnpm exec eslint --no-warn-ignored packages/react-on-rails/src/rendererTeardown.ts packages/react-on-rails/src/ClientRenderer.ts packages/react-on-rails/src/createReactOutput.ts packages/react-on-rails-pro/src/ClientSideRenderer.ts packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx packages/react-on-rails/tests/ClientRenderer.test.ts`
- `pnpm --filter react-on-rails run type-check`
- `pnpm --filter react-on-rails-pro run type-check`
- `pnpm --filter react-on-rails exec jest tests/ClientRenderer.test.ts --runInBand`
- `pnpm --filter react-on-rails-pro exec jest tests/ClientSideRenderer.test.ts tests/wrapServerComponentRenderer.client.test.jsx --runInBand`
- `pnpm --filter react-on-rails run build`
- `pnpm --filter react-on-rails-pro run build`
- Follow-up focused reruns after review-comment cleanup: `pnpm exec prettier --check packages/react-on-rails/tests/ClientRenderer.test.ts packages/react-on-rails-pro/tests/wrapServerComponentRenderer.client.test.jsx`, `pnpm exec eslint --no-warn-ignored packages/react-on-rails/tests/ClientRenderer.test.ts packages/react-on-rails-pro/tests/wrapServerComponentRenderer.client.test.jsx`, `pnpm --filter react-on-rails exec jest tests/ClientRenderer.test.ts --runInBand`, `pnpm --filter react-on-rails-pro exec jest tests/wrapServerComponentRenderer.client.test.jsx --runInBand`.

## Design Calls
- Added an internal OSS export instead of documenting three synchronized `isRendererTeardownResult` copies; Pro already consumes OSS internal subpaths, and this keeps the guard shape single-sourced.
- Kept `invokeRendererTeardown` duplicated; that helper still has package-specific runtime/logging context and remains covered by parity tests.
- Kept public `RendererFunction` return typing unchanged pending downstream migration data.
- Left docs and dummy-app renderer conversions out of this PR per scope; this PR stays focused on renderer teardown diagnostics, guard reuse, and targeted tests.
- No changelog entry: this is Pro/OSS renderer hardening plus an invalid-shape diagnostic guard, not a new user-facing feature or documented behavior change.

Labels: none. This is a focused test/diagnostic hardening PR with no CI workflow, dependency, generator, SSR output, or broad Pro/core boundary changes.

---

> [!NOTE]
> **Low Risk**
> Internal guard deduplication and stricter runtime validation for invalid server-component render shapes; behavior changes are limited to clearer errors and test-documented teardown edge cases, with no auth or data-path changes.
> 
> **Overview**
> **Centralizes** the `isRendererTeardownResult` type guard in `rendererTeardown.ts` and exposes it as `react-on-rails/@internal/rendererTeardown`, replacing inline copies in OSS (`ClientRenderer`, `createReactOutput`) and Pro (`ClientSideRenderer`, `wrapServerComponentRenderer/client`).
> 
> **Pro RSC wrapper** now rejects server-component render functions that resolve to a teardown object with an explicit error (before DOM/context checks), closing #3647-style misconfiguration diagnostics.
> 
> **Tests** add coverage for non-native thenable teardown rejections (Pro), same-id async teardown races where a stale renderer must not overwrite a newer mount (OSS), and `wrapServerComponentRenderer` validation for missing `domNodeId`, DOM node, `railsContext`, and teardown-shaped returns; webpack manifest test aliases the new internal module.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b92db44df38c91ceafc87c1b0ff13ca80c090108. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Added runtime validation that rejects invalid server-rendered results (e.g., teardown objects) with clearer error messages.
  * Improved handling and logging when asynchronous renderer teardown or replacement fails or races.
  * Improved validation and error messages for missing mount target or missing context during server-component mounting.

* **Tests**
  * Expanded tests for server component rendering validation edge cases.
  * Added tests ensuring async renderer teardown/replacement logs correctly for promise/thenable scenarios.

### CODE REVIEW

**Comment by justin808** [comment]
Pushed review follow-up commits:

- f0979f3e0d7db0099bf4c77908a931ddb2674ce8 centralizes isRendererTeardownResult in react-on-rails/@internal/rendererTeardown and updates OSS/Pro call sites to share it.
- 98c93ecff1c570bfcc218af02d313ffb14a06407 addresses the current test-readability/helper-cleanup threads.

Validation run locally:
- pnpm install --frozen-lockfile
- pnpm exec prettier --check on touched renderer teardown files/tests
- pnpm exec eslint --no-warn-ignored on touched renderer teardown files/tests
- pnpm --filter react-on-rails run type-check
- pnpm --filter react-on-rails-pro run type-check
- pnpm --filter react-on-rails exec jest tests/ClientRenderer.test.ts --runInBand
- pnpm --filter react-on-rails-pro exec jest tests/ClientSideRenderer.test.ts tests/wrapServerComponentRenderer.client.test.jsx --runInBand
- pnpm --filter react-on-rails run build
- pnpm --filter react-on-rails-pro run build

I’m resolving only the review threads covered by those commits.

File: packages/react-on-rails/tests/ClientRenderer.test.ts [hunk 1]
```diff
@@ -410,6 +368,48 @@ describe('ClientRenderer', () => {
...
+        // only the mount node for this id.
+        renderComponent('renderer-stale-after-newer');
+        expect(resolvers).toHaveLength(2);
+
+        const staleTeardown = jest.fn();
+        const currentTeardown = jest.fn();
+        // Resolve the newer renderer first so its teardown is attached to the active entry before
+        // the stale renderer settles.
+        resolvers[1]({ teardown: currentTeardown });
+        await Promise.resolve();
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/ClientRenderer.test.ts:400]: Deferring this nit from the current renderer teardown lane because this pass is scoped to the Pro wrapServerComponentRenderer tests and alias setup. Renderer teardown polish remains tracked in #3647.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 2]
```diff
@@ -75,6 +76,13 @@ const wrapServerComponentRenderer = (
       ? ((await componentOrRenderFunction(props, railsContext, domNodeId)) as ReactComponent)
       : componentOrRenderFunction;
 
+    if (isRendererTeardownResult(Component)) {
+      throw new Error(
+        `wrapServerComponentRenderer: render function for server component '${componentName}' ` +
+          'returned a renderer teardown result; expected a React component.',
+      );
+    }
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:87]: Deferred to follow-up #3647: hoist domNodeId / railsContext validation before awaiting render functions so invalid inputs are rejected before async renderer work.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:87]: Added a compatibility comment documenting that existing 3-arg render functions are awaited before domNodeId validation. Moving that validation earlier is intentionally deferred to #3647.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 3]
```diff
@@ -33,6 +38,14 @@ import handleRecoverableError from '../handleRecoverableError.client.ts';
 
 ensureReactUseAvailable();
 
+function isRendererTeardownResult(value: unknown): value is RendererTeardownResult {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx]: Superseded by the shared internal rendererTeardown helper: ClientRenderer and ClientSideRenderer now import isRendererTeardownResult instead of duplicating that type guard.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 4]
```diff
@@ -75,6 +76,15 @@ const wrapServerComponentRenderer = (
       ? ((await componentOrRenderFunction(props, railsContext, domNodeId)) as ReactComponent)
       : componentOrRenderFunction;
 
+    // Preserve compatibility with existing 3-arg render functions: they are awaited before domNodeId
+    // validation. Moving that guard earlier is intentionally deferred to #3647.
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx]: Added a comment documenting that the teardown-result guard intentionally runs before DOM/context checks so wrong-shaped render results get the clearer component-shape error.

File: packages/react-on-rails/tests/ClientRenderer.test.ts [hunk 5]
```diff
@@ -410,6 +368,49 @@ describe('ClientRenderer', () => {
...
+        resolvers[0]({ teardown: staleTeardown });
+        await Promise.resolve();
+
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/ClientRenderer.test.ts:403]: The current line anchor drifted, but I updated the relevant non-native thenable test comment to explain the timing: Promise.resolve(nonNativeThenable) first schedules thenable assimilation, and the downstream handler captures the teardown after that flush.

File: packages/react-on-rails/tests/ClientRenderer.test.ts [hunk 6]
```diff
@@ -410,6 +368,50 @@ describe('ClientRenderer', () => {
       consoleErrorSpy.mockRestore();
     });
 
+    it('does not let a stale async renderer overwrite a newer same-id teardown', async () => {
+      // No additional setupRailsContext() call needed; the enclosing beforeEach already runs it.
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/ClientRenderer.test.ts:372]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 7]
```diff
@@ -75,6 +76,16 @@ const wrapServerComponentRenderer = (
       ? ((await componentOrRenderFunction(props, railsContext, domNodeId)) as ReactComponent)
       : componentOrRenderFunction;
 
+    // Preserve compatibility with existing 3-arg render functions: they are awaited before domNodeId
+    // validation; moving that check earlier is a future improvement. Keep this teardown-result guard
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:80]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/tests/wrapServerComponentRenderer.client.test.jsx [hunk 8]
```diff
@@ -37,6 +37,86 @@ const createBailoutDigestError = () => {
...
+  }));
+
+  // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
+  const wrapServerComponentRenderer = require('../src/wrapServerComponentRenderer/client.tsx').default;
+
+  return { wrapServerComponentRenderer, createRoot, hydrateRoot, render, unmount, getReactServerComponent };
+};
+
+const cleanupWrappedRendererMocks = () => {
+  document.body.innerHTML = '';
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/wrapServerComponentRenderer.client.test.jsx:63]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts [hunk 9]
```diff
@@ -395,6 +395,42 @@ describe('ClientSideRenderer', () => {
...
+      ComponentRegistry.register({ TestComponent: TestRenderer });
+      const componentSpec = setupTestComponentDom('dom-id-teardown-thenable-reject');
+      addRailsContext();
+
+      await renderOrHydrateComponent(componentSpec);
+      unmountAll();
+      expect(teardown).toHaveBeenCalledTimes(1);
+
+      await Promise.resolve(); // lets Promise.resolve(nonNativeThenable) settle
+      await Promise.resolve(); // lets the .catch() rejection handler run
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts:424]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 10]
```diff
@@ -75,6 +76,16 @@ const wrapServerComponentRenderer = (
...
+    // before DOM checks so wrong-shaped render results get the clearer error.
+    if (isRendererTeardownResult(Component)) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:82]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails/tests/ClientRenderer.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/tests/ClientRenderer.test.ts:372]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts [hunk 11]
```diff
@@ -395,6 +395,42 @@ describe('ClientSideRenderer', () => {
...
+              onRejected(rejection);
+            },
+          }) as unknown as Promise<void>,
+      );
+      const TestRenderer: RendererFunction = (
+        _props?: Record<string, unknown>,
+        _railsContext?: RailsContext,
+        _domNodeId?: string,
+      ) => ({ teardown });
+      ComponentRegistry.register({ TestComponent: TestRenderer });
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/ClientSideRenderer.test.ts:415]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 12]
```diff
@@ -75,6 +76,16 @@ const wrapServerComponentRenderer = (
       ? ((await componentOrRenderFunction(props, railsContext, domNodeId)) as ReactComponent)
       : componentOrRenderFunction;
 
+    // Preserve compatibility with existing 3-arg render functions: they are awaited before domNodeId
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:79]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.

File: packages/react-on-rails/src/rendererTeardown.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import type { RendererTeardownResult } from './types/index.ts';
+
+// eslint-disable-next-line import/prefer-default-export -- single-export module; named export keeps the type guard's API tied to its predicate name.
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails/src/rendererTeardown.ts:3]: PR #3663 is already merged, so I am tracking this post-merge instead of changing the merged branch: #3675. Resolving this thread in favor of that follow-up issue.
